### PR TITLE
[SDK-3769] Respect rtl12

### LINKED ISSUE #1790: Library doesn't respect RTL12
[(RTL12)](https://sdk.ably.com/builds/ably/specification/main/features/#RTL12) An attached channel may receive an additional ATTACHED ProtocolMessage from Ably at any point. (This is typically triggered following a transport being resumed to indicate a partial loss of message continuity on that channel, in which case the ProtocolMessage will have a resumed flag set to false). **If and only if the resumed flag is false, this should result in the channel emitting an UPDATE event** with a ChannelStateChange object. The ChannelStateChange object should have both previous and current attributes set to attached, the reason attribute set to to the error member of the ATTACHED ProtocolMessage (if any), and the resumed attribute set per the RESUMED bitflag of the ATTACHED ProtocolMessage. (Note that UPDATE should be the only event emitted: in particular, the library must not emit an ATTACHED event if the channel was already attached, see RTL2g).

But library [emits](https://github.com/ably/ably-cocoa/blob/273277e566fc1641cc8fc5b6473999a3e9ab6c4e/Source/ARTRealtimeChannel.m#L748) an UPDATE event regardless said condition.

### PR DESCRIPTION
**Pull request by maratal**
Closes #1790

### CODE REVIEW

File: Test/Tests/RealtimeClientChannelTests.swift [hunk 1]
```diff
@@ -521,15 +522,12 @@ class RealtimeClientChannelTests: XCTestCase {
...
-
-                channel.on { stateChange in
-                    if stateChange.current == .attached {
-                        XCTAssertTrue(stateChange.resumed)
-                        XCTAssertNil(stateChange.reason)
-                        XCTAssertEqual(stateChange.current, ARTRealtimeChannelState.attached)
-                        XCTAssertEqual(stateChange.previous, ARTRealtimeChannelState.attached)
-                        done()
-                    }
+                XCTAssertEqual(channel.state, ARTRealtimeChannelState.attached)
```
Comment (lawrence-forooghian) [inline Test/Tests/RealtimeClientChannelTests.swift:525]: I don't understand the idea behind these test changes — why is it no longer making assertions about the `ATTACHED` `ChannelStateChange`, which is what RTL2f talks about? And how is it related to the above change, which relates to when an `UPDATE` `ChannelStateChange` will be emitted?

Comment (maratal) [reply Test/Tests/RealtimeClientChannelTests.swift:525]: Looks like RTL2f talks about `ChannelStateChange` missing details of `RTL12` which states that only in case of `resumed == false` `channel.on` will be triggered. This test checks the case when `resumed == true`, that's why I've removed `channel.on` (bc it will not be triggered). So the new assertions claim that channel states remains the same before and after ATTACHED protocol message and that it contains `resumed == true`.

Comment (lawrence-forooghian) [reply Test/Tests/RealtimeClientChannelTests.swift:525]: `RTL2f` is about what value to use for the `resumed` value of a `ChannelStateChange`. That means that this test is no longer testing `RTL2f` as it advertises, since as you've said there is no channel state change any more.

Comment (maratal) [reply Test/Tests/RealtimeClientChannelTests.swift]: I've [removed](https://github.com/ably/ably-cocoa/pull/1791/commits/1a5f933f4068398009cf0641bfdfb76844fa5baa) `RTL2f` test and created [issue](https://github.com/ably/ably-cocoa/issues/1812) to create a proper one.

Comment (maratal) [reply Test/Tests/RealtimeClientChannelTests.swift]: Also updated `RTL12` test - https://github.com/ably/ably-cocoa/pull/1791/commits/9184efa3eece85f60db12d0cf69cada8a6ba5319

File: Test/Tests/RealtimeClientChannelTests.swift [hunk 2]
```diff
@@ -3791,10 +3721,13 @@ class RealtimeClientChannelTests: XCTestCase {
             attachedMessageWithError.action = .attached
             attachedMessageWithError.channel = channel.name
 
-            channel.once(.update) { stateChange in
+            channel.on { stateChange in
```
Comment (lawrence-forooghian) [inline Test/Tests/RealtimeClientChannelTests.swift:3724]: Why this change from `once` to `on`?

**Review by lawrence-forooghian — COMMENTED** [review]
Which test changes reflect the changed behaviour of `ARTRealtimeChannel`? It's not clear to me.

File: Test/Tests/RealtimeClientChannelTests.swift (hunk 2 above)
Comment (lawrence-forooghian) [inline Test/Tests/RealtimeClientChannelTests.swift:3724]: Why this change from `once` to `on`?

File: Source/ARTRealtimeChannel.m [hunk 3]
```diff
@@ -741,11 +741,13 @@ - (void)setAttached:(ARTProtocolMessage *)message {
     }
 
     if (self.state_nosync == ARTRealtimeChannelAttached) {
-        if (message.error != nil) {
-            _errorReason = message.error;
+        if (!message.resumed) { // RTL12
```
Comment (lawrence-forooghian) [inline Source/ARTRealtimeChannel.m:744]: Where is this new behaviour tested? It's not clear to me.

File: Test/Tests/RealtimeClientChannelTests.swift (hunk 2 above)
Comment (maratal) [reply Test/Tests/RealtimeClientChannelTests.swift:3724]: I think it should be `on`, bc in case of invalid event the test will fail (on event assertion), which will signal about improper library behavior.

Comment (lawrence-forooghian) [reply Test/Tests/RealtimeClientChannelTests.swift:3724]: OK, makes sense.

File: Source/ARTRealtimeChannel.m (hunk 3 above)
Comment (maratal) [reply Source/ARTRealtimeChannel.m:744]: https://github.com/ably/ably-cocoa/pull/1791/commits/18f5f6ca27f47eaccb234d57a3a6350e03e504f6

**Review by lawrence-forooghian — APPROVED** [review]
